### ISSUE TITLE
False-positive `signature-differs` / `arguments-differ` with `typing.overload`

### ISSUE THREAD

**Issue by cdce8p** [PyCQA/pylint#5264]
### Bug description

Noticed that one while working on https://github.com/PyCQA/astroid/pull/1217 with @DanielNoord.

The logic for `signature-differs` / `arguments-differ` can't handle function redefinitions as its common with `typing.overload`. In particular I narrowed it down to two issue:
1. All function definitions seem to be check, even those decorated with `overload`.
2. The point of comparison is always the first definition, even if that isn't the true signature.

--
The examples require **Python 3.8** (due to the use of `typing.Literal`). Not that they aren'T necessarily typed correctly. Just for illustration proposes.

**Example 1**
```py
# pylint: disable=unused-argument,no-self-use,missing-docstring
from typing import Literal, Union, overload

class Parent:
    @overload
    def statement(self) -> int:
        ...

    @overload
    def statement(self, future: Literal[True]) -> str:
        ...

    def statement(self, future: Literal[None, True] = None) -> Union[int, str]:
        pass


class Child(Parent):
    # Signature comparison happens with first definition - the overload
    # Should use the actual function instead!
    def statement(self, future=None):  # arguments-differ (false-positive)
        pass
```

**Example 2**
```py
# pylint: disable=unused-argument,no-self-use,missing-docstring
from typing import Literal, Union, overload

class Parent2:
    def statement(self, future: Literal[None, True] = None) -> Union[int, str]:
        pass

class Child2(Parent2):
    @overload
    def statement(self, future: Literal[None] = ...) -> int:
        ...

    @overload
    def statement(self, future: Literal[True]) -> str:  # signature-differs (false-positive)
        # Even overloads are checked although these aren't required
        # to have the correct signature
        ...

    def statement(self, future: Literal[None, True] = None) -> Union[int, str]:
        pass
```

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
W0221: Number of parameters was 1 in 'Parent.statement' and is now 2 in overridden 'Child.statement' method (arguments-differ)

W0222: Signature differs from overridden 'statement' method (signature-differs)
```

### Expected behavior

_no errors_

### Pylint version

```shell
2.11.2-dev0
```

**Comment by DanielNoord** [comment]
@cdce8p I narrowed down the issue with selecting the first overload instead of the actual signature due to the `__getitem__` method of `LocalsDictNodeNG` in `astroid. See:
https://github.com/PyCQA/astroid/blob/e3730a7aaf35030edcafbb5ec0509fb8ff070e33/astroid/nodes/scoped_nodes.py#L316-L324

This returns `self.locals[item][0]`. For `overload` this doesn't make sense and should probably be `self.locals[item][-1]`. I wonder if any other checks actually need to use `[0]`: if there are multiple definitions of a method in a class definition won't the last one be used anyway?

Do we want to fix this in `astroid` by changing the `__getitem__` method (either by checking of decorated with overload, or changing for all methods) or do we want to create a workaround in `pylint`?

**Comment by Pierre-Sassoulas** [comment]
The first example was fixed in 2.15.0-dev0, not the second one.
